src/combinatorics/quiver/cast.lean

```diff
@@ -0,0 +1,114 @@
+/-
+Copyright (c) 2022 Antoine Labelle, Rémi Bottinelli. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Labelle, Rémi Bottinelli
+-/
+import combinatorics.quiver.basic
+import combinatorics.quiver.path
+
+/-!
+
+# Rewriting arrows and paths along vertex equalities
+
+This files defines `hom.cast` and `path.cast` (and associated lemmas) in order to allow
+rewriting arrows and paths along equalities of their endpoints.
+
+-/
+
+universes v v₁ v₂ u u₁ u₂
+
+variables {U : Type*} [quiver.{u+1} U]
+
+namespace quiver
+
+/-!
+### Rewriting arrows along equalities of vertices
+-/
+
+/-- Change the endpoints of an arrow using equalities. -/
+def hom.cast {u v u' v' : U} (hu : u = u') (hv : v = v') (e : u ⟶ v) : u' ⟶ v' :=
+eq.rec (eq.rec e hv) hu
+
+lemma hom.cast_eq_cast {u v u' v' : U} (hu : u = u') (hv : v = v') (e : u ⟶ v) :
+  e.cast hu hv = cast (by rw [hu, hv]) e :=
+eq.drec (eq.drec (eq.refl (hom.cast (eq.refl u) (eq.refl v) e)) hu) hv
+
+@[simp] lemma hom.cast_rfl_rfl {u v : U} (e : u ⟶ v) :
+  e.cast rfl rfl = e := rfl
+
+@[simp] lemma hom.cast_cast {u v u' v' u'' v'' : U} (e : u ⟶ v)
+  (hu : u = u') (hv : v = v') (hu' : u' = u'') (hv' : v' = v'') :
+  (e.cast hu hv).cast hu' hv' = e.cast (hu.trans hu') (hv.trans hv') :=
+by { subst_vars, refl }
+
+lemma hom.cast_heq {u v u' v' : U} (hu : u = u') (hv : v = v') (e : u ⟶ v) :
+  e.cast hu hv == e :=
+by { rw hom.cast_eq_cast, exact cast_heq _ _ }
+
+lemma hom.cast_eq_iff_heq {u v u' v' : U} (hu : u = u') (hv : v = v')
+  (e : u ⟶ v) (e' : u' ⟶ v') : e.cast hu hv = e' ↔ e == e' :=
+by { rw hom.cast_eq_cast, exact cast_eq_iff_heq }
+
+lemma hom.eq_cast_iff_heq {u v u' v' : U} (hu : u = u') (hv : v = v')
+  (e : u ⟶ v) (e' : u' ⟶ v') : e' = e.cast hu hv ↔ e' == e :=
+⟨λ h, ((e.cast_eq_iff_heq hu hv e').1 h.symm).symm,
+ λ h, ((e.cast_eq_iff_heq hu hv e').2 h.symm).symm⟩
```

Comment:
I feel like this would be clearer, but no need to change it.
```suggestion
by { rw [eq_comm, hom.cast_eq_iff_heq], exact ⟨heq.symm, heq.symm⟩ }
```
(That exact should be replaced by `rw heq_comm` if only `heq_comm` existed.)